File: aemet/SergioDeIscar-aemet-java/SergioDeIscar-aemet-java/src/main/java/org/example/services/storage/dupla/DuplaFileXml.java
package org.example.services.storage.dupla;

import org.example.config.AppConfig;
import org.example.dto.DuplaDto;
import org.example.dto.DuplasDto;
import org.example.models.Dupla;
import org.simpleframework.xml.core.Persister;

import java.io.File;
import java.io.IOException;
import java.util.List;

public class DuplaFileXml implements DuplaStorageService{
    //Singleton
    private static DuplaFileXml instance = null;
    public static DuplaFileXml getInstance() {
        if (instance == null) instance = new DuplaFileXml();
        return instance;
    }

    private final String localPath = AppConfig.getInstance().getAppData() + File.separator + "duplas.xml";

    @Override
    public List<Dupla> saveAll(List<Dupla> elements) {
        File file = new File(localPath);
        if (file.exists() && !file.canWrite() ) return null;
        Persister persister = new Persister();
        try {
            persister.write(new DuplasDto(
                elements.stream().map(Dupla::toDto).toList()
            ), file);
            return elements;
        }catch (Exception e) {
            System.out.println("Error al escribir en el fichero XML");
        }
        return null;
    }

    @Override
    public List<Dupla> loadAll() {
        File file = new File(localPath);
        if(!file.exists() || !file.canRead()) return null;
        Persister persister = new Persister();
        try{
            return persister.read(DuplasDto.class, file)
                .getDuplas()
                .stream()
                .map(DuplaDto::toClass)
                .toList();
        } catch (Exception e) {
            System.out.println("Error al leer en el fichero XML");
        }
        return null;
    }
}

File: aemet/SergioDeIscar-aemet-kotlin/SergioDeIscar-aemet-kotlin/src/main/kotlin/services/storage/dupla/DuplaFileXml.kt
package services.storage.dupla

import config.AppConfig
import dto.DuplasDto
import models.Dupla
import org.simpleframework.xml.core.Persister
import java.io.File

object DuplaFileXml: DuplaStorageService {

    private val localFile = "${AppConfig.APP_DATA}${File.separator}duplas.xml"

    override fun saveAll(elements: List<Dupla>): List<Dupla> {
        val file = File(localFile)
        if (file.exists() && !file.canWrite()) return emptyList()

        val listDto = elements.map { it.toDto() }
        val accidentesDto = DuplasDto(listDto)

        val serializer = Persister()
        serializer.write(accidentesDto, file)
        return elements
    }

    override fun loadAll(): List<Dupla> {
        val file = File(localFile)
        if (!file.exists() || !file.canRead()) return emptyList()
        val serializer = Persister()
        val accidentes = serializer
            .read(DuplasDto::class.java, file)
        return accidentes.duplas.map { it.toClass() }
    }
}